src/coreclr/vm/gcheaputilities.h

```diff
@@ -5,13 +5,65 @@
 #define _GCHEAPUTILITIES_H_
 
 #include "gcinterface.h"
+#include "math.h"
 
 // The singular heap instance.
 GPTR_DECL(IGCHeap, g_pGCHeap);
 
 #ifndef DACCESS_COMPILE
 extern "C" {
 #endif // !DACCESS_COMPILE
+
+// This struct adds some state that is only visible to the EE onto the standard gc_alloc_context
+typedef struct _ee_alloc_context
+{
+    // This is the address where the next allocation sampling should occur:
+    //  - if sampling is off, this is the same as alloc_limit
+    //  - if sampling is on, this is the next sampled byte in the gc_alloc_context
+    //  - if no such byte exists (i.e. the sampling threshold is outside of the allocation context),
+    //     this is the same as alloc_limit
+    uint8_t* alloc_sampling;
+    gc_alloc_context gc_alloc_context;
+
+    // TODO: how to get a random number?
+    // - have a dedicated CLRRandom instance: seems like a waste because the related Thread already has its CLRRandom instance
+    // - share the CLRRandom instance of the related thread but the Init() method of the allocation context runs BEFORE the thread's m_random.Init() is called
+    // Do we want to inject the Thread's CLRRandom instance into the allocation context?  Could be done in the Init() method of the Thread.InitThread() method
+    // CLRRandom* m_pRandom;
+    // Another solution would be to pass the CLRRandom instance as a parameter to the ComputeSamplingLimit() method
+
+ public:
+    void init()
+    {
+        LIMITED_METHOD_CONTRACT;
+
+        // we can't compute a sampling limit
+        // because we don't know the size of the allocation context yet
+        alloc_sampling = nullptr;
+        gc_alloc_context.init();
+    }
+
+    inline void ComputeSamplingLimit(CLRRandom* pRandomizer, size_t mean)
+    {
+        // TODO: maybe it is easier to assume that the caller of this function will check if sampling is on/off
+        // If sampling is off this is just setting alloc_sampling = alloc_limit
+        // If sampling is on then we'd do some pseudo-random number generation to decide what is
+        // the next sampled byte in the gc_alloc_context, if any.
+
+        //TODO: implement sampling limit placement strategy
+        double probability = pRandomizer->NextDouble();
+        size_t threshold = (size_t)(-log((1-probability) * mean) + 1);
```

Comment:
Off by one?
```suggestion
        size_t threshold = (size_t)-log((1-probability) * mean);
```
I assume:
 - threshold of 0 = sample the first byte
 - threshold of alloc_limit-1 = sample the last byte in the AC
 - threshold of >= alloc_limit = don't sample within this AC

Reply:
Definitively need to start at 0.

I always struggle with the limits: do you know if alloc_limit points to the last byte of the allocation context or to the byte after the allocation context?  I'm assuming it is the former.
The fast path assembly code checks that size + alloc_ptr > alloc_sampling before taking the slow path: it means that the last byte of the allocation context cannot be part of an allocated object:

          0   1   2   3
        |    |    |    |    |
ptr ---^            ^--- sampling
if size = 3, ptr + size = 3 that is equal to sampling so take the slow path

Anyway, I agree with your assumptions and I'm changing the code to remove the +1

Reply:
> I always struggle with the limits: do you know if alloc_limit points to the last byte of the allocation context or to the byte after the allocation context? I'm assuming it is the former.

It is the latter. As a concrete example if alloc_ptr = 1000 and alloc_limit = 1012 this alloc_context is just big enough to allocate a 12 byte object that resides in bytes [1000,1011]. The byte at address 1012 can not be used as part of an allocation.